refactor(core): Stop injecting global settings via angular and provide some typings around settings

jrsquared opened the pull request:
The global settings were passed around by injecting them with an angular constant. Since we're in the ES6 world, it makes more sense to import them.  `settings` was interleaved throughout much of deck and breaking it out into a standard ES6 module makes general refactoring much easier...
This still wraps the window.spinnakerSettings that the angular one wrapped, but in a much cleaner fashion.

Also in this is providing typings for the settings that I know about, while still satisfying typescript for fields that we do not know about yet. (by using `[key: string]: any` where applicable in `settings.ts`)

There's a lot of changes in here, but it's all mainly the same thing...
* _classes_: remove settings injection, add settings import
* _tests_: remove settings injection, add settings import, setup a default settings fallback for changes from tests.
* There were a few places we were injecting settings when we didn't need to as well, so removed those without replacing with an import.

@anotherchrisberry @zanthrash @icfantv @danielpeach PTAL

icfantv commented:
OMG.  I think I might love you.

danielpeach commented on app/scripts/modules/core/config/settings.ts:
```diff
@@ -0,0 +1,159 @@
+  };
+}
+
+interface IAppEngineProviderSettings {
```
I might not be 100% consistent, but I try to not camel case App Engine as one word, i.e., `Appengine`, since the API does the same.

jrsquared replied:
Cool, fixed.

jrsquared commented:
@danielpeach I thought about add helpers, but in that case it's just wrapping lodash functions and not adding any extra functionality around them. My end thought was to avoid the need to even use lodash for `get` or `has`, so we can still use types. If you are passionate about it, I'm happy to add the helper functions, but otherwise, I lean toward just leaving them out so it's more a conscious decision to not directly access properties.

icfantv commented:
@jrsquared, what if the helper functions gave lodash types?  Ala, `get<T>(foo): <T>` or something like that.  On phone so can't check exact syntax.

anotherchrisberry commented on app/scripts/modules/amazon/serverGroup/configure/serverGroupCommandBuilder.aws.service.spec.js:
```diff
@@ -10,17 +11,22 @@ describe('Service: awsServerGroup', function () {
+  const DEFAULT_SETTINGS = _.cloneDeep(SETTINGS);
+  afterEach(function () {
+    _.merge(SETTINGS, DEFAULT_SETTINGS);
+  });
```
arrow function (to make copy/pasting this a little cleaner)?
```javascript
afterEach(() => _.merge(SETTINGS, DEFAULT_SETTINGS));
```

jrsquared replied:
Yep, #facepalm

anotherchrisberry commented on app/scripts/modules/core/application/modal/applicationProviderFields.component.spec.js:
```diff
@@ -37,19 +30,23 @@ describe('Controller: ApplicationProviderFieldsCtrl', function () {
+  const DEFAULT_SETTINGS = _.cloneDeep(SETTINGS);
+  afterEach(function () {
+    _.merge(SETTINGS, DEFAULT_SETTINGS);
+    SETTINGS.providers.aws.defaultPath = '/path/to/somewhere';
```
This seems wrong - did you mean to put this in a `beforeEach` block?

jrsquared replied:
Yep, good catch. Tests still passed, but left SETTINGS wrong.

anotherchrisberry commented on app/scripts/modules/core/authentication/authentication.provider.spec.ts:
```diff
@@ -36,22 +39,23 @@ describe('authenticationProvider: application startup', function () {
   afterEach(function () {
-    settings.authEnabled = false;
+    merge(SETTINGS, DEFAULT_SETTINGS);
+    SETTINGS.authEnabled = false;
```
same - should this be in a `beforeEach`?

jrsquared replied:
Ah, right, I just left it in the afterEach from the refactor. I'll fix it.

anotherchrisberry commented on app/scripts/modules/core/config/settings.ts:
```diff
@@ -0,0 +1,174 @@
+    account: string;
+    editLoadBalancerStageEnabled: boolean;
+  };
+}
```
Not crazy about having provider-specific interfaces in the core module. Probably better to move these interfaces into the provider modules, then have a marker interface for `IProviderConfig` or something, with `IProviderSettings` below just being:
```
interface IProviderSettings {
  [key: string]: IProviderConfig
}
```

Does that make sense?

jrsquared replied:
Yep, makes perfect sense. Kept this at a 1:1 to current and was going to figure out provider specific later but makes sense to make that change now.

jrsquared replied:
Actually, I'm not entirely sure how to do this without forcing explicit casting anywhere the provider settings want to be typescript aware...

anotherchrisberry replied:
Yeah, it would involve an explicit cast, but that would keep provider code in provider modules.

jrsquared replied:
Alright, I agree with specific provider code in specific modules, so for now, instead of moving all the provider specific stuff out, I'll just remove that level of typing for now, since it really didn't matter, and we can revisit provider specific settings typings later.

anotherchrisberry commented on app/scripts/modules/core/config/settings.ts:
```diff
@@ -0,0 +1,174 @@
+    blesk?: boolean;
+    tableau?: boolean;
+    [key: string]: any;
+  };
```
Can this just be
```
feature {
  [key: string]: boolean
}
```
? Or do we have non-flags in the feature block?

jrsquared replied:
It can. But having all the ones we know about provides typeahead in IDEs, which is nice. If you don't think that's worth it, I can take it out.

anotherchrisberry replied:
Good point, I have no idea. I do like typeahead.

anotherchrisberry commented on app/scripts/modules/core/config/settings.ts:
```diff
@@ -0,0 +1,174 @@
+// Make sure to set up some reasonable default settings fields so we do not have to keep checking if they exist everywhere
+SETTINGS.feature = SETTINGS.feature || {};
+SETTINGS.analytics = SETTINGS.analytics || {};
+SETTINGS.providers = SETTINGS.providers || {};
```
It might be nice to add a couple of bits here to make test setup/cleanup easier:
```
const originalSettings: ISpinnakerSettings = _.cloneDeep(SETTINGS);
SETTINGS.afterTest = () => { _.merge(SETTINGS, originalSettings); }
```

Then, in tests, you can just call:
```javascript
afterEach(SETTINGS.afterTest);
```

jrsquared replied:
Ooo I like it.

jrsquared replied:
Done. Much better.

anotherchrisberry commented on app/scripts/modules/core/entityTag/entityTags.read.service.spec.ts:
```diff
@@ -10,37 +11,43 @@ describe('entityTags reader', () => {
-  let settings: any;
+
+  beforeEach(function () {
+    // Why do we have to clear the settings file? What setting is causing retries to not work?
```
really great questions.

anotherchrisberry commented:
Generally speaking, this is great.

jrsquared commented:
Alright, I added types for each providers settings, and those types only exist within each providers module.  There is a a helper function in each module settings to just call `SETTINGS.resetToOriginal` so there does not have to be another dependency on SETTINGS just to call the reset function.

@danielpeach I would appreciate another look so you can comment on the module-specific settings typings.

This also removed most (all?) of the get/has operators on settings, so eliminates the need for helper functions in settings too.

anotherchrisberry commented:
Very nice. LGTM, defer to @danielpeach

jrsquared commented:
Yep, the docker builds fine.

jrsquared commented on app/scripts/modules/core/config/settings.ts:
```diff
@@ -0,0 +1,113 @@
+  providers?: {
+    [key: string]: IProviderSettings; // allows custom providers not typed in here (good for testing too)
+  };
+  whatsNew: {
```
True... I'll look into if/how it makes sense to separate out these settings.
